### ISSUE TITLE
Training doesn't work on TF 2.3.1 version

### ISSUE THREAD

**Issue by hkang106** [xuannianz/EfficientDet#218]
It gives me this error message on TF 2.3.1:

valueerror: it appears you are trying to construct a functional model, but not all of the inputs in the first positional argument of your layer call are symbolic tensors. (input objects, or the output of another layer) functional models cannot correctly track custom layers unless all values in the first call argument are symbolic.

So I downgraded to TF 2.1, and it works fine.

**Comment by Ivandanilovich** [comment]
Problem is in ClassNet and BoxNet. I am trying to rewrite it for tf 2.3.*.
I am stuck that I absolutely don't understand what is going on here.
`def call(self, inputs, **kwargs):  
        feature, level = inputs  
        for i in range(self.depth):  
            feature = self.convs[i](feature)  
            feature = self.bns[i][self.level](feature)  
            feature = self.relu(feature) 
        outputs = self.head(feature)  
        outputs = self.reshape(outputs)  
        outputs = self.activation(outputs)
        self.level += 1
        return outputs`
**level** passed into **call** but never used. **self.level** are used as index for BatchNorm layers and always incremented. WTF? How it works?

**Comment by BonVant** [comment]
Yeah the idea behind it is to share the layers weights accross all levels **except** for the batch norm. But it is not very well put and in tf 2.3 if I recall correctly self.level will be incremented till **IndexError** raises: indeed __call__ is used also during training (triggering incrementation of self.level which is not intended behavior).
In general model subclassing is not on point and I would not recommand it. It leads to having to use the tf savedFormat for instance which currently doesn't support custom gradients (which are used here with custom losses).

Have a look at the way they handle it in keras-retinanet package using tf.keras.Model(inputs=inputs, outputs=outputs). You can still define layers in your BoxNet and ClassNet (but not subclassing tf.keras.Models) to share weights. And your __call__ should look like this:

```Python
  def __call__(self, shape, level):
        inputs = tf.keras.Input(shape=tuple(shape[1:]))
        for i in range(self.depth):
            features = self.convs[i](inputs)
            features = self.bns[i][level](features)
            features = self.relu(features)
        outputs = self.head(features)
        outputs = self.reshape(outputs)
        return tf.keras.Model(inputs=inputs, outputs=outputs, name=f"{self.name}_{level}")
```

with for BoxNet for instance :
```Python
box_net = BoxNet(*args, **kwargs)  # your params
regression = [box_net(feature.get_shape().as_list(), level)(feature) for level, feature in enumerate(fpn_features)]
```

**Comment by 518789Adf** [comment]
It appears you are trying to construct a functional model, but not all of the inputs in the first positional argument of your layer call are symbolic tensors. (Input objects, or the output of another layer) Functional models cannot correctly track custom layers unless all values in the first call argument are symbolic.
